<a href="https://colab.research.google.com/github/fralfaro/MAT306/blob/main/docs/labs/lab_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# MAT306 - Laboratorio N°03





**Objetivo**: Aplicar técnicas avanzadas de manipulación y análisis de datos con pandas sobre un conjunto real de datos de contenido de Netflix, reforzando buenas prácticas y métodos eficientes sin recurrir a `groupby`, `merge`, `pivot`, ni `join`.



**Dataset**:

Trabajaremos con el archivo `netflix_titles.csv`, que contiene información sobre los títulos disponibles en la plataforma Netflix hasta el año 2021.

| Variable       | Clase     | Descripción                                                                 |
|----------------|-----------|------------------------------------------------------------------------------|
| show_id        | caracter  | Identificador único del título en el catálogo de Netflix.                   |
| type           | caracter  | Tipo de contenido: 'Movie' o 'TV Show'.                                     |
| title          | caracter  | Título del contenido.                                                       |
| director       | caracter  | Nombre del director (puede ser nulo).                                       |
| cast           | caracter  | Lista de actores principales (puede ser nulo).                              |
| country        | caracter  | País o países donde se produjo el contenido.                                |
| date_added     | fecha     | Fecha en la que el título fue agregado al catálogo de Netflix.              |
| release_year   | entero    | Año de lanzamiento original del título.                                     |
| rating         | caracter  | Clasificación por edad (por ejemplo: 'PG-13', 'TV-MA').                      |
| duration       | caracter  | Duración del contenido (minutos o número de temporadas para series).        |
| listed_in      | caracter  | Categorías o géneros en los que está clasificado el contenido.              |
| description    | caracter  | Breve sinopsis del contenido.                                               |




In [50]:
import pandas as pd

# Cargar datos
df = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT306/main/docs/labs/data/netflix_titles.csv')
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...


### Parte 1: Limpieza y preparación

1. Revisar y describir el dataset:

   * ¿Cuántas filas y columnas tiene?
   * ¿Qué tipos de datos hay?
   * ¿Cuántos valores nulos hay por columna?

2. Transformar la columna `date_added` a tipo fecha.

3. Crear columnas auxiliares con `assign`:

   * Año (`year_added`)
   * Mes (`month_added`)



In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [52]:
df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

Hay 8807 filas y 12 columnas, se tienen datos enteros y tipo object.

En la siguiente tabla se ven la cantidad de valores nulos por columna

In [53]:
# valores nulos
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [54]:
import datetime
# Eliminar espacios al inicio y al final
df['date_added'] = df['date_added'].str.strip()

df['date_added'] = pd.to_datetime(df['date_added'], format='%B %d, %Y')
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   
3 2021-09-24          2021  TV-MA   1 Season   
4 2021-09-24          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [55]:
df = df.assign(year_added = df['date_added'].dt.year)
df = df.assign(month_added = df['date_added'].dt.month)
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   
3 2021-09-24          2021  TV-MA   1 Season   
4 2021-09-24          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  year_added  month_added  
0  As her father nears the end of his life, filmm...      2021.0          9.0  
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0  
2  To protect his family from a powerful drug lor...      2021.0          9.0  
3  Feuds, flirtations and toilet talk go down amo...      2021.0          9.0  
4  In a city of coaching centers known to train I...      2021.0          9.0

## Parte 2: Técnicas avanzadas de pandas

4. Utilizar `.loc` para seleccionar películas (`type == 'Movie'`) que fueron agregadas después del año 2018.

5. Utilizar `str.contains()` y `str.extract()`:

   * Filtrar títulos que contienen la palabra 'love' (sin distinguir mayúsculas/minúsculas).
   * Extraer la duración en minutos para las películas desde la columna `duration`.

6. Aplicar `explode()` sobre la columna `listed_in` para obtener una fila por cada género.

7. Obtener un top 10 de géneros más frecuentes utilizando `value_counts()`.

8. Aplicar `where()` y `mask()` para marcar las películas de más de 120 minutos como contenido largo en una nueva columna.

9. Utilizar `.loc` para filtrar películas que cumplen con:

   * Más de 100 minutos de duración.
   * Rating igual a `'R'`.
   * País igual a `'United States'`.

10. Utilizar `.style` para formatear visualmente el top 10 de películas más largas.

In [56]:
#4
df_pelicula = df.loc[(df['type'] == 'Movie') & (df['year_added'] > 2018)]

df_pelicula.head()

show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country date_added  \
0                                       United States 2021-09-25   
6                                                 NaN 2021-09-24   
7   United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                       United States 2021-09-24   
12                            Germany, Czech Republic 2021-09-23   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  year_added  month_added  
0   As her father nears the end of his life, filmm...      2021.0          9.0  
6   Equestria's divided. But a bright-eyed hero be...      2021.0          9.0  
7   On a photo shoot in Ghana, an American model s...      2021.0          9.0  
9   A woman adjusting to life after a loss contend...      2021.0          9.0  
12  After most of her family is murdered in a terr...      2021.0          9.0

In [57]:
#5
#Copiamos a un nuevo dataframe los datos donde el titulo contiene la palabra love
df_love = df[df['title'].str.contains("love", case=False)]
df_love[['title', 'type']].head()

title     type
25        Love on the Spectrum  TV Show
158    Love Don't Cost a Thing    Movie
159             Love in a Puff    Movie
206  LSD: Love, Sex Aur Dhokha    Movie
227                Really Love    Movie

In [58]:
#Añadimoos una nueva columna donde se añaden los minutos de duracion, filtrado para que solo corresponda a las peliculas
df['duracion_min'] = df.loc[df['type'] == "Movie", 'duration'].str.extract(r'(\d+)').astype(float)
df.head(10)

show_id     type                             title  \
0      s1    Movie              Dick Johnson Is Dead   
1      s2  TV Show                     Blood & Water   
2      s3  TV Show                         Ganglands   
3      s4  TV Show             Jailbirds New Orleans   
4      s5  TV Show                      Kota Factory   
5      s6  TV Show                     Midnight Mass   
6      s7    Movie  My Little Pony: A New Generation   
7      s8    Movie                           Sankofa   
8      s9  TV Show     The Great British Baking Show   
9     s10    Movie                      The Starling   

                        director  \
0                Kirsten Johnson   
1                            NaN   
2                Julien Leclercq   
3                            NaN   
4                            NaN   
5                  Mike Flanagan   
6  Robert Cullen, José Luis Ucha   
7                   Haile Gerima   
8                Andy Devonshire   
9                 Theodore Melfi   

                                                cast  \
0                                                NaN   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
3                                                NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...   
6  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8  Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   

                                             country date_added  release_year  \
0                                      United States 2021-09-25          2020   
1                                       South Africa 2021-09-24          2021   
2                                                NaN 2021-09-24          2021   
3                                                NaN 2021-09-24          2021   
4                                              India 2021-09-24          2021   
5                                                NaN 2021-09-24          2021   
6                                                NaN 2021-09-24          2021   
7  United States, Ghana, Burkina Faso, United Kin... 2021-09-24          1993   
8                                     United Kingdom 2021-09-24          2021   
9                                      United States 2021-09-24          2021   

  rating   duration                                          listed_in  \
0  PG-13     90 min                                      Documentaries   
1  TV-MA  2 Seasons    International TV Shows, TV Dramas, TV Mysteries   
2  TV-MA   1 Season  Crime TV Shows, International TV Shows, TV Act...   
3  TV-MA   1 Season                             Docuseries, Reality TV   
4  TV-MA  2 Seasons  International TV Shows, Romantic TV Shows, TV ...   
5  TV-MA   1 Season                 TV Dramas, TV Horror, TV Mysteries   
6     PG     91 min                           Children & Family Movies   
7  TV-MA    125 min   Dramas, Independent Movies, International Movies   
8  TV-14  9 Seasons                       British TV Shows, Reality TV   
9  PG-13    104 min                                   Comedies, Dramas   

                                         description  year_added  month_added  \
0  As her father nears the end of his life, filmm...      2021.0          9.0   
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0   
2  To protect his family from a powerful drug lor...      2021.0          9.0   
3  Feuds, flirtations and toilet talk go down amo...      2021.0          9.0   
4  In a city of coaching centers known to train I...      2021.0          9.0   
5  The arrival of a charismatic young priest brin...      2021.0          9.0   
6  Equestria's divided. But a bright-eyed hero be...      2021.0          9.0   
7  On a photo sh

In [59]:
#6
#Separamos por coma
df['listed_in'] = df['listed_in'].str.split(',')

df_exploded = df.explode('listed_in')

df_exploded.head(15)


show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
1      s2  TV Show          Blood & Water              NaN   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
2      s3  TV Show              Ganglands  Julien Leclercq   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   
4      s5  TV Show           Kota Factory              NaN   
4      s5  TV Show           Kota Factory              NaN   
5      s6  TV Show          Midnight Mass    Mike Flanagan   
5      s6  TV Show          Midnight Mass    Mike Flanagan   
5      s6  TV Show          Midnight Mass    Mike Flanagan   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...            NaN   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...            NaN   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...            NaN   

  date_added  release_year rating   duration                listed_in  \
0 2021-09-25          2020  PG-13     90 min            Documentaries   
1 2021-09-24          2021  TV-MA  2 Seasons   International TV Shows   
1 2021-09-24          2021  TV-MA  2 Seasons                TV Dramas   
1 2021-09-24          2021  TV-MA  2 Seasons             TV Mysteries   
2 2021-09-24          2021  TV-MA   1 Season           Crime TV Shows   
2 2021-09-24          2021  TV-MA   1 Season   International TV Shows   
2 2021-09-24          2021  TV-MA   1 Season    TV Action & Adventure   
3 2021-09-24          2021  TV-MA   1 Season               Docuseries   
3 2021-09-24          2021  TV-MA   1 Season               Reality TV   
4 2021-09-24          2021  TV-MA  2 Seasons   International TV Shows   
4 2021-09-24          2021  TV-MA  2 Seasons        Romantic TV Shows   
4 2021-09-24          2021  TV-MA  2 Seasons              TV Comedies   
5 2021-09-24          2021  TV-MA   1 Season                TV Dramas   
5 2021-09-24          2021  TV-MA   1 Season                TV Horror   
5 2021-09-24          2021  TV-MA   1 Season             TV Mysteries   

                                         description  year_added  month_added  \
0  As her father nears the end of his life, filmm...      2021.0          9.0   
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0   
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0   
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0   
2  To protect his family from a powerful drug lor...      2021.0          9.0   
2  To protect his family from a powerful drug lor...      2021.0          9.0   
2  To protect his family from a powerful drug lor...      2021.0          9.0   
3  Feuds, flirtations and toilet talk go down amo...   

In [60]:
#7
df_exploded['listed_in'].value_counts()

listed_in
 International Movies        2624
Dramas                       1600
Comedies                     1210
Action & Adventure            859
Documentaries                 829
                             ... 
Romantic Movies                 3
Spanish-Language TV Shows       2
LGBTQ Movies                    1
TV Sci-Fi & Fantasy             1
Sports Movies                   1
Name: count, Length: 73, dtype: int64

El genero mas repetido es "International Movies"

In [65]:
#8
df['Cantidad Contenido'] = "Normal"
df['Cantidad Contenido'] = df['Cantidad Contenido'].mask(df['duracion_min'] > 120, "Contenido Largo")

df.head(15)

show_id     type                                title  \
0       s1    Movie                 Dick Johnson Is Dead   
1       s2  TV Show                        Blood & Water   
2       s3  TV Show                            Ganglands   
3       s4  TV Show                Jailbirds New Orleans   
4       s5  TV Show                         Kota Factory   
5       s6  TV Show                        Midnight Mass   
6       s7    Movie     My Little Pony: A New Generation   
7       s8    Movie                              Sankofa   
8       s9  TV Show        The Great British Baking Show   
9      s10    Movie                         The Starling   
10     s11  TV Show  Vendetta: Truth, Lies and The Mafia   
11     s12  TV Show                     Bangkok Breaking   
12     s13    Movie                         Je Suis Karl   
13     s14    Movie     Confessions of an Invisible Girl   
14     s15  TV Show      Crime Stories: India Detectives   

                         director  \
0                 Kirsten Johnson   
1                             NaN   
2                 Julien Leclercq   
3                             NaN   
4                             NaN   
5                   Mike Flanagan   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
8                 Andy Devonshire   
9                  Theodore Melfi   
10                            NaN   
11              Kongkiat Komesiri   
12            Christian Schwochow   
13                  Bruno Garotti   
14                            NaN   

                                                 cast  \
0                                                 NaN   
1   Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
2   Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
3                                                 NaN   
4   Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
5   Kate Siegel, Zach Gilford, Hamish Linklater, H...   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8   Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
10                                                NaN   
11  Sukollawat Kanarot, Sushar Manaying, Pavarit M...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
13  Klara Castanho, Lucca Picon, Júlia Gomes, Marc...   
14                                                NaN   

                                              country date_added  \
0                                       United States 2021-09-25   
1                                        South Africa 2021-09-24   
2                                                 NaN 2021-09-24   
3                                                 NaN 2021-09-24   
4                                               India 2021-09-24   
5                                                 NaN 2021-09-24   
6                                                 NaN 2021-09-24   
7   United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
8                                      United Kingdom 2021-09-24   
9                                       United States 2021-09-24   
10                                                NaN 2021-09-24   
11                                                NaN 2021-09-23   
12                            Germany, Czech Republic 2021-09-23   
13                                                NaN 2021-09-22   
14                                                NaN 2021-09-22   

    release_year rating   duration  \
0           2020  PG-13     90 min   
1           2021  TV-MA  2 Seasons   
2           2021  TV-MA   1 Season   
3           2021  TV-MA   1 Season   
4           2021  TV-MA  2 Seasons   
5           2021  TV-MA   1 Season   
6           2021     PG     91 min   
7           1993  TV-MA    125 min   
8           2021  TV-14  9 Seasons   
9           2021  PG-13    104 min   
10          2021  TV-MA   1 S

In [67]:

df_pelis100R = df.loc[(df['type'] == 'Movie') & (df['duracion_min'] > 100) & (df['rating'] == 'R') &(df['country'] == 'United States')]

df_pelis100R.head()

show_id   type                        title               director  \
48      s49  Movie                 Training Day          Antoine Fuqua   
81      s82  Movie                         Kate  Cedric Nicolas-Troyan   
131    s132  Movie  Blade Runner: The Final Cut           Ridley Scott   
139    s140  Movie           Do the Right Thing              Spike Lee   
144    s145  Movie                  House Party        Reginald Hudlin   

                                                  cast        country  \
48   Denzel Washington, Ethan Hawke, Scott Glenn, T...  United States   
81   Mary Elizabeth Winstead, Jun Kunimura, Woody H...  United States   
131  Harrison Ford, Rutger Hauer, Sean Young, Edwar...  United States   
139  Danny Aiello, Ossie Davis, Ruby Dee, Richard E...  United States   
144  Christopher Reid, Christopher Martin, Robin Ha...  United States   

    date_added  release_year rating duration  \
48  2021-09-16          2001      R  122 min   
81  2021-09-10          2021      R  106 min   
131 2021-09-01          1982      R  117 min   
139 2021-09-01          1989      R  120 min   
144 2021-09-01          1990      R  104 min   

                                             listed_in  \
48                                [Dramas,  Thrillers]   
81                                [Action & Adventure]   
131  [Action & Adventure,  Classic Movies,  Cult Mo...   
139               [Classic Movies,  Comedies,  Dramas]   
144                           [Comedies,  Cult Movies]   

                                           description  year_added  \
48   A rookie cop with one day to prove himself to ...      2021.0   
81   Slipped a fatal poison on her final job, a rut...      2021.0   
131  In a smog-choked dystopian Los Angeles, blade ...      2021.0   
139  On a sweltering day in Brooklyn, simmering rac...      2021.0   
144  Grounded by his strict father, Kid risks life ...      2021.0   

     month_added  duracion_min Cantidad Contenido  
48           9.0         122.0    Contenido Largo  
81           9.0         106.0             Normal  
131          9.0         117.0             Normal  
139          9.0         120.0             Normal  
144          9.0         104.0             Normal

In [71]:
#10
top10 = df.loc[df['type'] == 'Movie'].nlargest(10, 'duracion_min')

top10[['title', 'duracion_min']].style.background_gradient(subset=['duracion_min'], cmap="Greens").set_caption("Top 10 de películas más largas")




### Pregunta Desafío

11. ¿Cuáles son las combinaciones más frecuentes de género y rating en el dataset?
    (Sugerencia: utilizar `value_counts` con `subset=["genre", "rating"]` después de aplicar `explode()`).



### Bonus: Análisis de duplicados y limpieza

12. ¿Existen películas con el mismo nombre (`title`) pero con distinto año de lanzamiento (`release_year`)?
13. ¿Cuántos títulos únicos hay en total en la columna `title`?





In [73]:
#11

comb_frec = (df_exploded.value_counts(subset=['listed_in', 'rating']).head(20))
print(comb_frec)


listed_in                rating
 International Movies    TV-MA     1074
                         TV-14     1022
Dramas                   TV-MA      616
                         TV-14      428
 TV Dramas               TV-MA      401
Comedies                 TV-MA      400
                         TV-14      393
International TV Shows   TV-MA      368
 International TV Shows  TV-MA      346
 Independent Movies      TV-MA      335
International TV Shows   TV-14      319
Documentaries            TV-MA      315
Crime TV Shows           TV-MA      304
Stand-Up Comedy          TV-MA      284
 International Movies    TV-PG      280
 Dramas                  TV-14      265
Dramas                   R          248
 TV Dramas               TV-14      240
 Thrillers               TV-MA      233
Action & Adventure       R          220
Name: count, dtype: int64
